flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/sink/writer/FileWriter.java

```diff
@@ -0,0 +1,278 @@
+	private final BucketAssigner<IN, BucketID> bucketAssigner;
+
+	private final BucketWriter<IN, BucketID> bucketWriter;
+
+	private final RollingPolicy<IN, BucketID> rollingPolicy;
+
+	// --------------------------- runtime fields -----------------------------
+
+	private final BucketerContext bucketerContext;
+
+	private final Map<BucketID, FileWriterBucket<IN, BucketID>> activeBuckets;
+
+	private final OutputFileConfig outputFileConfig;
+
+	// --------------------------- State Related Fields -----------------------------
+
+	private final FileWriterBucketStateSerializer<BucketID> bucketStateSerializer;
+
+	/**
+	 * A constructor creating a new empty bucket manager.
+	 *
+	 * @param basePath The base path for our buckets.
+	 * @param bucketAssigner The {@link BucketAssigner} provided by the user.
+	 * @param bucketFactory The {@link FileWriterBucketFactory} to be used to create buckets.
+	 * @param bucketWriter The {@link BucketWriter} to be used when writing data.
+	 * @param rollingPolicy The {@link RollingPolicy} as specified by the user.
+	 */
+	public FileWriter(
+			final Path basePath,
+			final BucketAssigner<IN, BucketID> bucketAssigner,
+			final FileWriterBucketFactory<IN, BucketID> bucketFactory,
+			final BucketWriter<IN, BucketID> bucketWriter,
+			final RollingPolicy<IN, BucketID> rollingPolicy,
+			final OutputFileConfig outputFileConfig) {
+
+		this.basePath = checkNotNull(basePath);
+		this.bucketAssigner = checkNotNull(bucketAssigner);
+		this.bucketFactory = checkNotNull(bucketFactory);
+		this.bucketWriter = checkNotNull(bucketWriter);
+		this.rollingPolicy = checkNotNull(rollingPolicy);
+
+		this.outputFileConfig = checkNotNull(outputFileConfig);
+
+		this.activeBuckets = new HashMap<>();
+		this.bucketerContext = new BucketerContext();
+
+		this.bucketStateSerializer = new FileWriterBucketStateSerializer<>(
+				bucketWriter.getProperties().getInProgressFileRecoverableSerializer(),
+				bucketAssigner.getSerializer());
+	}
+
+	/**
+	 * Initializes the state after recovery from a failure.
+	 *
+	 * <p>During this process:
+	 * <ol>
+	 *     <li>we set the initial value for part counter to the maximum value used before across all tasks and buckets.
+	 *     This guarantees that we do not overwrite valid data,</li>
+	 *     <li>we commit any pending files for previous checkpoints (previous to the last successful one from which we restore),</li>
+	 *     <li>we resume writing to the previous in-progress file of each bucket, and</li>
+	 *     <li>if we receive multiple states for the same bucket, we merge them.</li>
+	 * </ol>
+	 *
+	 * @param bucketStates the state holding recovered state about active buckets.
+	 *
+	 * @throws Exception if anything goes wrong during retrieving the state or restoring/committing of any
+	 * 		in-progress/pending part files
+	 */
+	public void initializeState(List<FileWriterBucketState<BucketID>> bucketStates) throws IOException {
+		for (FileWriterBucketState<BucketID> state : bucketStates) {
```

Comment:
I think we should add `checkState(bucketStates != null, "The retrieved state was null.");` or sth like that here.